### PR TITLE
new Fibonacci number rule

### PR DESCRIPTION
**Pull request by PanteraLX**
# Fibonacci

 - `v::arrayType()`

Validates whether the input follows the fibonacci integer sequence

```php
v::fibonacci()->validate(1); // true
v::fibonacci()->validate('34'); // true
v::fibonacci()->validate(6); // false
```

### CODE REVIEW

**Comment by PanteraLX** [comment]
I'll look into integration tests if I have some time this weekend

File: docs/Fibonacci.md [hunk 1]
```diff
@@ -0,0 +1,17 @@
+# Fibonacci
+
+ - `v::arrayType()`
```
Comment (henriquemoody) [inline docs/Fibonacci.md]: Fibonacci()

File: docs/Fibonacci.md [hunk 2]
```diff
@@ -0,0 +1,17 @@
...
+
+Validates whether the input follows the fibonacci integer sequence
```
Comment (henriquemoody) [inline docs/Fibonacci.md]: fibonacci must always starts with F in upper case, also you forget the dot (.) in the end of this paragraph.

File: docs/VALIDATORS.md [hunk 3]
```diff
@@ -52,6 +52,7 @@
   * [BoolType](BoolType.md)
   * [Even](Even.md)
   * [Factor](Factor.md)
+  * [Fibonacci](Fibonacci.md)
```
Comment (henriquemoody) [inline docs/VALIDATORS.md]: You should also add it to the _Alphabetically_ group.

File: library/Exceptions/FibonacciException.php [hunk 4]
```diff
@@ -0,0 +1,24 @@
...
+ * file that was distributed with this source code.
+ */
+
+namespace Respect\Validation\Exceptions;
+
+class FibonacciException extends ValidationException
+{
+    public static $defaultTemplates = [
+        self::MODE_DEFAULT => [
+            self::STANDARD => '{{name}} must be a valid fibonacci number',
```
Comment (henriquemoody) [inline library/Exceptions/FibonacciException.php]: Same I already [commented](https://github.com/Respect/Validation/pull/637/files#r50696141)

File: library/Exceptions/FibonacciException.php [hunk 5]
```diff
@@ -0,0 +1,24 @@
...
+        ],
+        self::MODE_NEGATIVE => [
+            self::STANDARD => '{{name}} must not be a valid fibonacci number',
```
Comment (henriquemoody) [inline library/Exceptions/FibonacciException.php]: Same I already [commented](https://github.com/Respect/Validation/pull/637/files#r50696141).

File: library/Exceptions/FibonacciException.php [hunk 6]
```diff
@@ -0,0 +1,24 @@
...
+        ],
+    ];
+}
```
Comment (henriquemoody) [inline library/Exceptions/FibonacciException.php]: All files must end with a new line at the end.

File: library/Rules/Fibonacci.php [hunk 7]
```diff
@@ -0,0 +1,34 @@
...
+        $fib[1] = 1;
+        $n = 1;
+        while ($input > $fib[$n]) {
+            $n++;
+            $fib[$n] = $fib[$n - 1] + $fib[$n - 2];
+        }
+
+        return $fib[$n] === (int)$input && is_numeric($input);
+    }
+}
```
Comment (henriquemoody) [inline library/Rules/Fibonacci.php]: Same I already [commented](https://github.com/Respect/Validation/pull/637/files#r50696393).

File: tests/unit/Rules/FibonacciTest.php [hunk 8]
```diff
@@ -0,0 +1,83 @@
...
+            [-1],
+            [5.2],
+            ['-1'],
+            ['a'],
+            [' '],
+            [false],
+            [true]
+        ];
+    }
+}
```
Comment (henriquemoody) [inline tests/unit/Rules/FibonacciTest.php]: Same I already [commented](https://github.com/Respect/Validation/pull/637/files#r50696393).

File: library/Rules/Fibonacci.php [hunk 9]
```diff
@@ -0,0 +1,34 @@
...
+ */
+class Fibonacci extends AbstractRule
+{
+    /**
+     * {@inheritdoc}
+     */
+    public function validate($input)
+    {
+        $fib[0] = 0;
+        $fib[1] = 1;
```
Comment (henriquemoody) [inline library/Rules/Fibonacci.php]: 1. Avoid abbreviations, `$sequence` would be a better name, IMO.
2. Always declare a variable before you use it, so you should not use `$fib[0] = 0` until you declare `$fib = []`.

In this case, I think `$sequence = [0, 1]` would be better.

File: library/Rules/Fibonacci.php [hunk 10]
```diff
@@ -0,0 +1,34 @@
...
+        $n = 1;
+        while ($input > $fib[$n]) {
+            $n++;
+            $fib[$n] = $fib[$n - 1] + $fib[$n - 2];
+        }
+
+        return $fib[$n] === (int)$input && is_numeric($input);
```
Comment (henriquemoody) [inline library/Rules/Fibonacci.php]: Just start the method with `filter_var($input, FILTER_VALIDATE_INT)`, then this validation is going to be simpler.

File: tests/unit/Rules/FibonacciTest.php [hunk 11]
```diff
@@ -0,0 +1,83 @@
...
+ */
+
+namespace Respect\Validation\Rules;
+
+/**
+ * @group  rule
+ * @covers Respect\Validation\Rules\Fibonacci
+ * @covers Respect\Validation\Exceptions\FibonacciException
+ */
+class FibonacciTest extends \PHPUnit_Framework_TestCase
```
Comment (henriquemoody) [inline tests/unit/Rules/FibonacciTest.php]: This test case can be a lot simpler, take a look on [ArrayTypeTest](https://github.com/Respect/Validation/blob/c2eece6/tests/unit/Rules/ArrayTypeTest.php).

**Comment by PanteraLX** [comment]
@henriquemoody Thank you for the feddback.

**Comment by henriquemoody** [comment]
Merged manually by ce3f88586280a27ecd730d542ae0f8a5c135edea

Thanks for contributing! :panda_face:

**Pull request closed without merging**